File: python/ader/weno.py
""" Implements the WENO method used in Dumbser et al (DOI 10.1016/j.cma.2013.09.022)
"""
from itertools import product

from numpy import ceil, concatenate, dot, floor, int64, multiply, ones, tensordot, zeros
from scipy.linalg import solve

from options import rc, λc, λs, eps, ndim, N, N1
from ader.basis import mid_values, end_values, derivative_end_values
from ader.weno_matrices import coefficient_matrices, oscillation_indicator
from gpr.variables.vectors import Cvec_to_Pvec


Mc = coefficient_matrices()
Σ = oscillation_indicator()
midvals = mid_values()


if N%2:
    nStencils = 4
    lamList = [λc, λc, λs, λs]
else:
    nStencils = 3
    lamList = [λc, λs, λs]


def extend(inarray, extension, d):
    """ Extends the input array by M cells on each surface
    """
    n = inarray.shape[d]
    reps = concatenate(([extension+1], ones(n-2), [extension+1])).astype(int64)
    return inarray.repeat(reps, axis=d)

def coeffs(uList):
    """ Calculate coefficients of basis polynomials and weights
    """
    wList = [solve(Mc[i], uList[i], overwrite_b=1, check_finite=0) for i in range(nStencils)]
    σList = [((w.T).dot(Σ).dot(w)).diagonal() for w in wList]
    oList = [lamList[i]  / (abs(σList[i]) + eps)**rc for i in range(nStencils)]
    oSum = zeros(18)
    numerator = zeros([N1, 18])
    for i in range(nStencils):
        oSum += oList[i]
        numerator += multiply(wList[i], oList[i])
    return numerator / oSum

def weno(u):
    """ Find reconstruction coefficients of u to order N+1
    """
    nx, ny, nz = u.shape[:3]
    floorHalfN = int(floor(N/2))
    ceilHalfN = int(ceil(N/2))

    Wx = zeros([nx, ny, nz, N1, 18])
    tempu = extend(u, N, 0)
    for i, j, k in product(range(nx), range(ny), range(nz)):
        ii = i + N
        if nStencils==3:
            u1 = tempu[ii-floorHalfN : ii+floorHalfN+1, j, k]
            u2 = tempu[ii-N : ii+1, j, k]
            u3 = tempu[ii : ii+N+1, j, k]
            Wx[i, j, k] = coeffs([u1, u2, u3])
        else:
            u1 = tempu[ii-floorHalfN : ii+ceilHalfN+1, j, k]
            u2 = tempu[ii-ceilHalfN : ii+floorHalfN+1, j, k]
            u3 = tempu[ii-N : ii+1, j, k]
            u4 = tempu[ii : ii+N+1, j, k]
            Wx[i, j, k] = coeffs([u1, u2, u3, u4])
    if ndim==1:
        return Wx

    Wxy = zeros([nx, ny, nz, N1, N1, 18])
    tempWx = extend(Wx, N, 1)
    for i, j, k in product(range(nx), range(ny), range(nz)):
        jj = j + N
        for a in range(N1):
            if nStencils==3:
                w1 = tempWx[i, jj-floorHalfN : jj+floorHalfN+1, k, a]
                w2 = tempWx[i, jj-N : jj+1, k, a]
                w3 = tempWx[i, jj : jj+N+1, k, a]
                Wxy[i, j, k, a] = coeffs([w1, w2, w3])
            else:
                w1 = tempWx[i, jj-floorHalfN : jj+ceilHalfN+1, k, a]
                w2 = tempWx[i, jj-ceilHalfN : jj+floorHalfN+1, k, a]
                w3 = tempWx[i, jj-N : jj+1, k, a]
                w4 = tempWx[i, jj : jj+N+1, k, a]
                Wxy[i, j, k, a] = coeffs([w1, w2, w3, w4])
    if ndim==2:
        return Wxy

    Wxyz = zeros([nx, ny, nz, N1, N1, N1, 18])
    tempWxy = extend(Wxy, N, 2)
    for i, j, k in product(range(nx), range(ny), range(nz)):
        kk = k + N
        for a, b in product(range(N1), range(N1)):
            if nStencils==3:
                w1 = tempWxy[i, j, kk-floorHalfN : kk+floorHalfN+1, a, b]
                w2 = tempWxy[i, j, kk-N : kk+1, a, b]
                w3 = tempWxy[i, j, kk : kk+N+1, a, b]
                Wxyz[i, j, k, a, b] = coeffs([w1, w2, w3])
            else:
                w1 = tempWxy[i, j, kk-floorHalfN : kk+ceilHalfN+1, a, b]
                w2 = tempWxy[i, j, kk-ceilHalfN : kk+floorHalfN+1, a, b]
                w3 = tempWxy[i, j, kk-N : kk+1, a, b]
                w4 = tempWxy[i, j, kk : kk+N+1, a, b]
                Wxyz[i, j, k, a, b] = coeffs([w1, w2, w3, w4])
    if ndim==3:
        return Wxyz

def weno_primitive(q, PAR, SYS):
    """ Returns a WENO reconstruction in primitive variables, given the grid of conserved values.
        A reconstruction in conserved variables is performed. The midpoints of this reconstruction
        are then taken as the conserved cell averages (required for >2nd order). A primitive
        reconstruction is then performed with these averages.
    """
    qr = weno(q)
    qav = dot(midvals, qr)
    pav = zeros(qav.shape)
    nx, ny, nz = qav.shape[:3]
    for i, j, k in product(range(nx), range(ny), range(nz)):
        pav[i,j,k] = Cvec_to_Pvec(qav[i,j,k], PAR, SYS)
    return weno(pav)

def weno_endpoints(wh):
    """ Returns an array containing the values of the basis polynomials at 0 and 1
        qEnd[d,e,i,j,k,:,:,:,:,:] is the set of coefficients at end e in the dth direction
    """
    endVals = end_values()
    derEndVals = derivative_end_values()
    nx, ny, nz = wh.shape[:3]
    wh0 = wh.reshape([nx, ny, nz] + [N1]*ndim + [18])
    wEnd = zeros([ndim, 2, nx, ny, nz] + [N1]*(ndim-1) + [18])
    wDerEnd = zeros([ndim, 2, nx, ny, nz] + [N1]*(ndim-1) + [18])
    for d in range(ndim):
        wEnd[d] = tensordot(endVals, wh0, (0,3+d))
        wDerEnd[d] = tensordot(derEndVals, wh0, (0,3+d))
    return wEnd, wDerEnd

File: julia/weno.jl
# Implements the WENO method used in Dumbser et al (DOI 10.1016/j.cma.2013.09.022)
include("weno_matrices.jl")

function extend(arr, N, nx, ny, nz, nvar, d)
  # Extends the input array by M cells on each surface
  if d==1
    ret = zeros(nx+2N, ny, nz, nvar)
    ret[N+1:nx+N, :, :, :] = arr
    in0 = arr[1, :, :, :]
    in1 = arr[nx, :, :, :]
    for i = 1:N
      ret[i, :, :, :] = in0
      ret[nx+N+i, :, :, :] = in1
    end
    return ret

  elseif d==2
    ret = zeros(nx, ny+2N, nz, N+1, nvar)
    ret[:, N+1:ny+N, :, :, :] = arr
    in0 = arr[:, 1, :, :]
    in1 = arr[:, ny, :, :]
    for i = 1:N
      ret[:, i, :, :] = in0
      ret[:, ny+N+i, :, :] = in1
    end
    return ret

  else
    ret = zeros(nx, ny, nz+2N, N+1, N+1, nvar)
    ret[:, :, N+1:nz+N, :, :, :] = arr
    in0 = arr[:, :, 1, :]
    in1 = arr[:, :, nz, :]
    for i = 1:N
      ret[:, :, i, :] = in0
      ret[:, :, nz+N+i, :] = in1
    end
    return ret
  end
end

@inline function indicator(chΣT, w, n)
  # The indicator is w' * Σ * w, ie |w' * chol(Σ)'|^2
  # NB The Cholesky decomposition always exists as Σ is SPD
  ret = 0.
  for j=1:n
    tmp = 0.
    for i=j:n
      @inbounds tmp += w[i] * chΣT[i,j]
    end
    ret += tmp^2
  end
  return ret
end

function coeffs3(ret, N, w1,w2,w3, M1,M2,M3, λs,λc, chΣT, ε, r)
  # Calculate coefficients of basis polynomials and weights
  A_ldiv_B!(M1, w1)
  A_ldiv_B!(M2, w2)
  A_ldiv_B!(M3, w3)
  o1 = λs / (indicator(chΣT, w1, N+1) + ε)^r
  o2 = λs / (indicator(chΣT, w2, N+1) + ε)^r
  o3 = λc / (indicator(chΣT, w3, N+1) + ε)^r
  oSum = o1 + o2 + o3
  for i = 1:N+1
    @inbounds ret[i] = (o1*w1[i] + o2*w2[i] + o3*w3[i]) / oSum
  end
end

function coeffs4(ret, N, w1,w2,w3,w4, M1,M2,M3,M4, λs,λc, chΣT, ε, r)
  # Calculate coefficients of basis polynomials and weights
  A_ldiv_B!(M1, w1)
  A_ldiv_B!(M2, w2)
  A_ldiv_B!(M3, w3)
  A_ldiv_B!(M4, w4)
  o1 = λs / (indicator(chΣT, w1, N+1) + ε)^r
  o2 = λs / (indicator(chΣT, w2, N+1) + ε)^r
  o3 = λc / (indicator(chΣT, w3, N+1) + ε)^r
  o4 = λc / (indicator(chΣT, w4, N+1) + ε)^r
  oSum = o1 + o2 + o3 + o4
  for i = 1:N+1
    @inbounds ret[i] = (o1*w1[i] + o2*w2[i] + o3*w3[i] + o4*w4[i]) / oSum
  end
end

function weno(u, N, M1, M2, M3, M4, chΣT, λc=1e5, λs=1, r=8, ε=1e-14)
  # Find reconstruction coefficients of u to order N+1
  nx, ny, nz, nvar = size(u)
  fHalfN = Int(floor(N/2))
  cHalfN = Int(ceil(N/2))
  n = Bool(N%2)

  w1 = Array{Float64}(N+1)
  w2 = Array{Float64}(N+1)
  w3 = Array{Float64}(N+1)
  w4 = Array{Float64}(N+1)

  Wx = zeros(nx, ny, nz, N+1, nvar)
  Wx0 = extend(u, N, nx, ny, nz, nvar, 1)

  for v=1:nvar, k=1:nz, j=1:ny
    w0 = view(Wx0, :, j, k, v)
    for i=1:nx
      for ind = 0:N
        @inbounds w1[ind+1] = w0[i+ind]
        @inbounds w2[ind+1] = w0[i+N+ind]
        @inbounds w3[ind+1] = w0[i+cHalfN+ind]
      end
      if n
        for ind = 0:N
          @inbounds w4[ind+1] = w0[i+fHalfN+ind]
        end
        coeffs4(view(Wx,i,j,k,:,v), N, w1,w2,w3,w4, M1,M2,M3,M4, λs,λc, chΣT,ε,r)
      else
        coeffs3(view(Wx,i,j,k,:,v), N, w1,w2,w3, M1,M2,M3, λs,λc, chΣT,ε,r)
      end
    end
  end
  if ny==1 && nz==1
    return Wx
  end

  Wxy = zeros(nx, ny, nz, N+1, N+1, nvar)
  Wxy0 = extend(Wx, N, nx, ny, nz, nvar, 2)

  for v=1:nvar, a=1:N+1, k=1:nz, i=1:nx
    w0 = view(Wxy0, i, :, k, a, v)
    for j=1:ny
      for ind = 0:N
        @inbounds w1[ind+1] = w0[j+ind]
        @inbounds w2[ind+1] = w0[j+N+ind]
        @inbounds w3[ind+1] = w0[j+cHalfN+ind]
      end
      if n
        for ind = 0:N
          @inbounds w4[ind+1] = w0[j+fHalfN+ind]
        end
        coeffs4(view(Wxy,i,j,k,a,:,v), N, w1,w2,w3,w4, M1,M2,M3,M4, λs,λc, chΣT,ε,r)
      else
        coeffs3(view(Wxy,i,j,k,a,:,v), N, w1,w2,w3, M1,M2,M3, λs,λc, chΣT,ε,r)
      end
    end
  end
  if nz==1
    return Wxy
  end

  Wxyz = zeros(nx, ny, nz, N+1, N+1, N+1, nvar)
  Wxyz0 = extend(Wxy, N, nx, ny, nz, nvar, 3)

  for v=1:nvar, b=1:N+1, a=1:N+1, j=1:ny, i=1:nx
    w0 = view(Wxyz0, i, j, :, a, b, v)
    for k=1:nz
      for ind = 0:N
        @inbounds w1[ind+1] = w0[k+ind]
        @inbounds w2[ind+1] = w0[k+N+ind]
        @inbounds w3[ind+1] = w0[k+cHalfN+ind]
      end
      if n
        for ind = 0:N
          @inbounds w4[ind+1] = w0[k+fHalfN+ind]
        end
        coeffs4(view(Wxyz,i,j,k,a,b,:,v), N, w1,w2,w3,w4, M1,M2,M3,M4, λs,λc, chΣT,ε,r)
      else
        coeffs3(view(Wxyz,i,j,k,a,b,:,v), N, w1,w2,w3, M1,M2,M3, λs,λc, chΣT,ε,r)
      end
    end
  end
  return Wxyz

end

function main(u, N)
  n = Bool(N%2)
  Mlist = coefficient_matrices(N, n)
  Σ = oscillation_indicator(N)
  M1 = factorize(Mlist[1])
  M2 = factorize(Mlist[2])
  M3 = factorize(Mlist[3])
  M4 = factorize(Mlist[4])
  chΣT = chol(Σ)'
  @time weno(u, N, M1, M2, M3, M4, chΣT)
end

function test()
  include("plot.jl")
  u = rand(91,1,1,18)
  u[:,1,1,1] = sin(1:0.1:10)
  u[:,1,1,2] = cos(1:0.1:10)
  test(u,2)
  wh = test(u,2)
  plot_weno(wh,1,2)
end
